client/tests/integration_test.rs

```diff
@@ -0,0 +1,160 @@
+use std::convert::Infallible;
+
+use futures::{channel::oneshot, future};
+use hyper::{Body, Response, Server};
+use serde::Deserialize;
+use tokio::task::JoinHandle;
+use warp::{Filter, Rejection};
+use warp_json_rpc::Builder;
+
+use casper_client::{DeployStrParams, PaymentStrParams, SessionStrParams};
+use casper_node::rpcs::{
+    account::{PutDeploy, PutDeployParams},
+    chain::{GetStateRootHash, GetStateRootHashParams},
+    info::{GetDeploy, GetDeployParams},
+    RpcWithOptionalParams, RpcWithParams,
+};
+
+fn test_filter<P>(
+    method: &'static str,
+) -> impl Filter<Extract = (Response<Body>,), Error = Rejection> + Copy
+where
+    for<'de> P: Deserialize<'de> + Send,
+{
+    warp_json_rpc::filters::json_rpc()
+        .and(warp_json_rpc::filters::method(method))
+        .and(warp_json_rpc::filters::params::<P>())
+        .map(|builder: Builder, _params: P| builder.success(()).unwrap())
+}
+
+struct MockServerHandle {
+    graceful_shutdown: oneshot::Sender<()>,
+    server_joiner: JoinHandle<Result<(), hyper::error::Error>>,
+}
+
+impl MockServerHandle {
+    async fn kill(self) {
+        let _ = self.graceful_shutdown.send(());
+        let _ = self.server_joiner.await;
+    }
+}
+
+/// Will spawn a server on localhost at `port` and respond to JSON-RPC requests that successfully
+/// deserialize as `P`.
+fn spawn_mock_rpc_server<P>(port: u16, method: &'static str) -> MockServerHandle
+where
+    P: 'static,
+    for<'de> P: Deserialize<'de> + Send,
+{
+    let service = warp_json_rpc::service(test_filter::<P>(method));
+    let builder = Server::try_bind(&([127, 0, 0, 1], port).into()).unwrap();
+    let make_svc =
+        hyper::service::make_service_fn(move |_| future::ok::<_, Infallible>(service.clone()));
+    let (graceful_shutdown, shutdown_receiver) = oneshot::channel::<()>();
+    let server = builder.serve(make_svc);
+    let server_with_shutdown = server.with_graceful_shutdown(async {
+        shutdown_receiver.await.ok();
+    });
+    let server_joiner = tokio::spawn(server_with_shutdown);
+    MockServerHandle {
+        graceful_shutdown,
+        server_joiner,
+    }
+}
+
+macro_rules! assert_valid_response {
```

Comment:
Can we avoid using a macro for this?  I think macros should be a last resort.

Reply:
Sure, removed this, it wasn't as compelling a use case as I thought it could be.